File: src/panel_order/layout.py
from __future__ import annotations

from dataclasses import dataclass

from .types import LayoutNode, Panel, Rect


@dataclass(slots=True)
class LayoutConfig:
    """Grouping and splitting thresholds for layout analysis."""

    min_split_gap_px: int = 12
    row_overlap_tolerance: float = 0.2
    column_overlap_tolerance: float = 0.2


class MangaLayoutAnalyzer:
    """Build a recursive layout tree and derive Japanese reading order."""

    def __init__(self, config: LayoutConfig | None = None) -> None:
        """Create a layout analyzer with the provided grouping configuration."""
        self.config = config or LayoutConfig()

    def build_layout(self, panels: list[Panel]) -> tuple[LayoutNode, list[int]]:
        """Build the layout tree and flatten it into reading order."""
        if not panels:
            raise ValueError("At least one panel is required to build a layout tree.")
        root = self._build_recursive(panels)
        return root, self._flatten_reading_order(root)

    def _build_recursive(self, panels: list[Panel]) -> LayoutNode:
        """Recursively partition panels into a tree of rows, columns, and leaves."""
        bounds = self._union_bounds([panel.rect for panel in panels])
        if len(panels) == 1:
            return LayoutNode(kind="leaf", bounds=bounds, panel_id=panels[0].panel_id)

        # Stage 1: Prefer explicit XY-cut splits when a clear empty gutter exists.
        split = self._best_xy_cut(panels)
        if split is not None:
            kind, first_group, second_group = split
            children = [self._build_recursive(first_group), self._build_recursive(second_group)]
            return LayoutNode(kind=kind, bounds=bounds, children=children)

        # Stage 2: Fall back to row grouping for top-to-bottom reading progression.
        rows = self._group_into_rows(panels)
        if len(rows) > 1:
            children = [self._build_recursive(row) for row in rows]
            return LayoutNode(kind="horizontal", bounds=bounds, children=children)

        # Stage 3: If rows fail, fall back to column grouping for right-to-left progression.
        columns = self._group_into_columns(panels)
        if len(columns) > 1:
            children = [self._build_recursive(column) for column in columns]
            return LayoutNode(kind="vertical", bounds=bounds, children=children)

        # Stage 4: Use a stable fallback order so every panel can still be serialized.
        fallback = sorted(panels, key=lambda panel: (panel.rect.center_y, -panel.rect.center_x))
        children = [self._build_recursive([panel]) for panel in fallback]
        return LayoutNode(kind="horizontal", bounds=bounds, children=children)

    def _best_xy_cut(self, panels: list[Panel]) -> tuple[str, list[Panel], list[Panel]] | None:
        """Choose the strongest empty-gutter split between vertical and horizontal cuts."""
        vertical = self._best_vertical_split(panels)
        horizontal = self._best_horizontal_split(panels)
        if vertical is None and horizontal is None:
            return None
        if vertical is None:
            _score, top_group, bottom_group = horizontal
            return "horizontal", top_group, bottom_group
        if horizontal is None:
            _score, left_group, right_group = vertical
            return "vertical", right_group, left_group
        if vertical[0] >= horizontal[0]:
            _score, left_group, right_group = vertical
            return "vertical", right_group, left_group
        _score, top_group, bottom_group = horizontal
        return "horizontal", top_group, bottom_group

    def _best_vertical_split(self, panels: list[Panel]) -> tuple[float, list[Panel], list[Panel]] | None:
        """Find the widest valid vertical gutter between two panel groups."""
        ordered = sorted(panels, key=lambda panel: panel.rect.x)
        prefix_max_right: list[int] = []
        suffix_min_left: list[int] = [0] * len(ordered)

        # Precompute occupied extents so every possible split can be scored in linear time.
        current_max = ordered[0].rect.right
        for panel in ordered:
            current_max = max(current_max, panel.rect.right)
            prefix_max_right.append(current_max)

        current_min = ordered[-1].rect.x
        for index in range(len(ordered) - 1, -1, -1):
            current_min = min(current_min, ordered[index].rect.x)
            suffix_min_left[index] = current_min

        best: tuple[float, list[Panel], list[Panel]] | None = None
        for index in range(len(ordered) - 1):
            # A valid vertical cut needs a visible empty gap between the left and right groups.
            gap = suffix_min_left[index + 1] - prefix_max_right[index]
            if gap < self.config.min_split_gap_px:
                continue
            score = float(gap)
            group_a = ordered[: index + 1]
            group_b = ordered[index + 1 :]
            if best is None or score > best[0]:
                best = (score, group_a, group_b)
        return best

    def _best_horizontal_split(self, panels: list[Panel]) -> tuple[float, list[Panel], list[Panel]] | None:
        """Find the widest valid horizontal gutter between two panel groups."""
        ordered = sorted(panels, key=lambda panel: panel.rect.y)
        prefix_max_bottom: list[int] = []
        suffix_min_top: list[int] = [0] * len(ordered)

        # Precompute occupied extents so every possible split can be scored in linear time.
        current_max = ordered[0].rect.bottom
        for panel in ordered:
            current_max = max(current_max, panel.rect.bottom)
            prefix_max_bottom.append(current_max)

        current_min = ordered[-1].rect.y
        for index in range(len(ordered) - 1, -1, -1):
            current_min = min(current_min, ordered[index].rect.y)
            suffix_min_top[index] = current_min

        best: tuple[float, list[Panel], list[Panel]] | None = None
        for index in range(len(ordered) - 1):
            # A valid horizontal cut needs a visible empty gap between the top and bottom groups.
            gap = suffix_min_top[index + 1] - prefix_max_bottom[index]
            if gap < self.config.min_split_gap_px:
                continue
            score = float(gap)
            group_a = ordered[: index + 1]
            group_b = ordered[index + 1 :]
            if best is None or score > best[0]:
                best = (score, group_a, group_b)
        return best

    def _group_into_rows(self, panels: list[Panel]) -> list[list[Panel]]:
        """Group panels into reading rows using vertical overlap."""
        ordered = sorted(panels, key=lambda panel: panel.rect.y)
        rows: list[list[Panel]] = []
        current: list[Panel] = []
        current_bottom = -1

        for panel in ordered:
            if not current:
                current = [panel]
                current_bottom = panel.rect.bottom
                continue

            # Panels that share enough vertical overlap belong to the same reading row.
            overlap = current_bottom - panel.rect.y
            min_height = min(min(item.rect.height for item in current), panel.rect.height)
            if overlap >= min_height * self.config.row_overlap_tolerance:
                current.append(panel)
                current_bottom = max(current_bottom, panel.rect.bottom)
                continue

            # Inside a row, Japanese manga panels are read from right to left.
            rows.append(sorted(current, key=lambda item: -item.rect.x))
            current = [panel]
            current_bottom = panel.rect.bottom

        if current:
            rows.append(sorted(current, key=lambda item: -item.rect.x))
        return rows

    def _group_into_columns(self, panels: list[Panel]) -> list[list[Panel]]:
        """Group panels into columns using horizontal overlap."""
        ordered = sorted(panels, key=lambda panel: panel.rect.x, reverse=True)
        columns: list[list[Panel]] = []
        current: list[Panel] = []
        current_left = 10**9

        for panel in ordered:
            if not current:
                current = [panel]
                current_left = panel.rect.x
                current_right = panel.rect.right
                continue

            # Panels that share enough horizontal overlap belong to the same column.
            overlap = panel.rect.right - current_left
            min_width = min(min(item.rect.width for item in current), panel.rect.width)
            if overlap >= min_width * self.config.column_overlap_tolerance:
                current.append(panel)
                current_left = min(current_left, panel.rect.x)
                current_right = max(current_right, panel.rect.right)
                continue

            columns.append(sorted(current, key=lambda item: item.rect.y))
            current = [panel]
            current_left = panel.rect.x
            current_right = panel.rect.right

        if current:
            columns.append(sorted(current, key=lambda item: item.rect.y))
        return columns

    def _flatten_reading_order(self, node: LayoutNode) -> list[int]:
        """Flatten the layout tree into the final reading order sequence."""
        if node.kind == "leaf":
            if node.panel_id is None:
                raise ValueError("Leaf node is missing panel_id.")
            return [node.panel_id]
        ordered: list[int] = []
        for child in node.children:
            ordered.extend(self._flatten_reading_order(child))
        return ordered

    def _union_bounds(self, rects: list[Rect]) -> Rect:
        """Return a single rectangle covering every rectangle in the input list."""
        bounds = rects[0]
        for rect in rects[1:]:
            bounds = bounds.union(rect)
        return bounds

File: web/src/core/panel_order/layout.ts
import type { PanelBox, PanelLayoutNode, PanelRect } from "./types";

type LayoutConfig = {
  minSplitGapPx: number;
  rowOverlapTolerance: number;
  columnOverlapTolerance: number;
};

const DEFAULT_LAYOUT_CONFIG: LayoutConfig = {
  minSplitGapPx: 12,
  rowOverlapTolerance: 0.2,
  columnOverlapTolerance: 0.2,
};

function rectRight(rect: PanelRect): number {
  return rect.x + rect.width;
}

function rectBottom(rect: PanelRect): number {
  return rect.y + rect.height;
}

function unionBounds(rects: PanelRect[]): PanelRect {
  const left = Math.min(...rects.map((rect) => rect.x));
  const top = Math.min(...rects.map((rect) => rect.y));
  const right = Math.max(...rects.map((rect) => rectRight(rect)));
  const bottom = Math.max(...rects.map((rect) => rectBottom(rect)));
  return { x: left, y: top, width: right - left, height: bottom - top };
}

function flattenReadingOrder(node: PanelLayoutNode): number[] {
  if (node.kind === "leaf") {
    if (node.panelId === undefined) {
      throw new Error("Leaf layout node is missing panelId.");
    }
    return [node.panelId];
  }
  return node.children.flatMap((child) => flattenReadingOrder(child));
}

function bestVerticalSplit(
  panels: PanelBox[],
  config: LayoutConfig,
): { score: number; firstGroup: PanelBox[]; secondGroup: PanelBox[] } | null {
  const ordered = [...panels].sort((left, right) => left.rect.x - right.rect.x);
  const prefixMaxRight: number[] = [];
  const suffixMinLeft: number[] = new Array(ordered.length);

  let currentMax = rectRight(ordered[0].rect);
  for (const panel of ordered) {
    currentMax = Math.max(currentMax, rectRight(panel.rect));
    prefixMaxRight.push(currentMax);
  }

  let currentMin = ordered[ordered.length - 1].rect.x;
  for (let index = ordered.length - 1; index >= 0; index -= 1) {
    currentMin = Math.min(currentMin, ordered[index].rect.x);
    suffixMinLeft[index] = currentMin;
  }

  let best: { score: number; firstGroup: PanelBox[]; secondGroup: PanelBox[] } | null = null;
  for (let index = 0; index < ordered.length - 1; index += 1) {
    const gap = suffixMinLeft[index + 1] - prefixMaxRight[index];
    if (gap < config.minSplitGapPx) {
      continue;
    }
    if (!best || gap > best.score) {
      best = {
        score: gap,
        firstGroup: ordered.slice(0, index + 1),
        secondGroup: ordered.slice(index + 1),
      };
    }
  }
  return best;
}

function bestHorizontalSplit(
  panels: PanelBox[],
  config: LayoutConfig,
): { score: number; firstGroup: PanelBox[]; secondGroup: PanelBox[] } | null {
  const ordered = [...panels].sort((left, right) => left.rect.y - right.rect.y);
  const prefixMaxBottom: number[] = [];
  const suffixMinTop: number[] = new Array(ordered.length);

  let currentMax = rectBottom(ordered[0].rect);
  for (const panel of ordered) {
    currentMax = Math.max(currentMax, rectBottom(panel.rect));
    prefixMaxBottom.push(currentMax);
  }

  let currentMin = ordered[ordered.length - 1].rect.y;
  for (let index = ordered.length - 1; index >= 0; index -= 1) {
    currentMin = Math.min(currentMin, ordered[index].rect.y);
    suffixMinTop[index] = currentMin;
  }

  let best: { score: number; firstGroup: PanelBox[]; secondGroup: PanelBox[] } | null = null;
  for (let index = 0; index < ordered.length - 1; index += 1) {
    const gap = suffixMinTop[index + 1] - prefixMaxBottom[index];
    if (gap < config.minSplitGapPx) {
      continue;
    }
    if (!best || gap > best.score) {
      best = {
        score: gap,
        firstGroup: ordered.slice(0, index + 1),
        secondGroup: ordered.slice(index + 1),
      };
    }
  }
  return best;
}

function groupIntoRows(panels: PanelBox[], config: LayoutConfig): PanelBox[][] {
  const ordered = [...panels].sort((left, right) => left.rect.y - right.rect.y);
  const rows: PanelBox[][] = [];
  let current: PanelBox[] = [];
  let currentBottom = -1;

  for (const panel of ordered) {
    if (current.length === 0) {
      current = [panel];
      currentBottom = rectBottom(panel.rect);
      continue;
    }

    const overlap = currentBottom - panel.rect.y;
    const minHeight = Math.min(...current.map((item) => item.rect.height), panel.rect.height);
    if (overlap >= minHeight * config.rowOverlapTolerance) {
      current.push(panel);
      currentBottom = Math.max(currentBottom, rectBottom(panel.rect));
      continue;
    }

    rows.push([...current].sort((left, right) => right.rect.x - left.rect.x));
    current = [panel];
    currentBottom = rectBottom(panel.rect);
  }

  if (current.length > 0) {
    rows.push([...current].sort((left, right) => right.rect.x - left.rect.x));
  }
  return rows;
}

function groupIntoColumns(panels: PanelBox[], config: LayoutConfig): PanelBox[][] {
  const ordered = [...panels].sort((left, right) => right.rect.x - left.rect.x);
  const columns: PanelBox[][] = [];
  let current: PanelBox[] = [];
  let currentLeft = Number.POSITIVE_INFINITY;

  for (const panel of ordered) {
    if (current.length === 0) {
      current = [panel];
      currentLeft = panel.rect.x;
      continue;
    }

    const overlap = rectRight(panel.rect) - currentLeft;
    const minWidth = Math.min(...current.map((item) => item.rect.width), panel.rect.width);
    if (overlap >= minWidth * config.columnOverlapTolerance) {
      current.push(panel);
      currentLeft = Math.min(currentLeft, panel.rect.x);
      continue;
    }

    columns.push([...current].sort((left, right) => left.rect.y - right.rect.y));
    current = [panel];
    currentLeft = panel.rect.x;
  }

  if (current.length > 0) {
    columns.push([...current].sort((left, right) => left.rect.y - right.rect.y));
  }
  return columns;
}

function buildRecursive(panels: PanelBox[], config: LayoutConfig): PanelLayoutNode {
  const bounds = unionBounds(panels.map((panel) => panel.rect));
  if (panels.length === 1) {
    return {
      kind: "leaf",
      bounds,
      panelId: panels[0].panelId,
      children: [],
    };
  }

  const vertical = bestVerticalSplit(panels, config);
  const horizontal = bestHorizontalSplit(panels, config);
  if (vertical || horizontal) {
    if (!horizontal || (vertical && vertical.score >= horizontal.score)) {
      return {
        kind: "vertical",
        bounds,
        // Japanese manga reads right group before left group.
        children: [buildRecursive(vertical!.secondGroup, config), buildRecursive(vertical!.firstGroup, config)],
      };
    }
    return {
      kind: "horizontal",
      bounds,
      children: [buildRecursive(horizontal.firstGroup, config), buildRecursive(horizontal.secondGroup, config)],
    };
  }

  const rows = groupIntoRows(panels, config);
  if (rows.length > 1) {
    return {
      kind: "horizontal",
      bounds,
      children: rows.map((row) => buildRecursive(row, config)),
    };
  }

  const columns = groupIntoColumns(panels, config);
  if (columns.length > 1) {
    return {
      kind: "vertical",
      bounds,
      children: columns.map((column) => buildRecursive(column, config)),
    };
  }

  const fallback = [...panels].sort((left, right) => {
    if (left.rect.y !== right.rect.y) {
      return left.rect.y - right.rect.y;
    }
    return right.rect.x - left.rect.x;
  });
  return {
    kind: "horizontal",
    bounds,
    children: fallback.map((panel) => buildRecursive([panel], config)),
  };
}

export function buildPanelLayout(
  panels: PanelBox[],
  config: Partial<LayoutConfig> = {},
): { layoutTree: PanelLayoutNode; readingOrder: number[] } {
  if (panels.length === 0) {
    throw new Error("At least one panel is required to build a layout tree.");
  }
  const resolvedConfig: LayoutConfig = { ...DEFAULT_LAYOUT_CONFIG, ...config };
  const layoutTree = buildRecursive(panels, resolvedConfig);
  return {
    layoutTree,
    readingOrder: flattenReadingOrder(layoutTree),
  };
}
